### DISCUSSION TITLE
fstab rootfs / switch_root

### DISCUSSION THREAD

**Discussion by pwFoo** [finit-project/finit#292] [Q&A] [answered]
Hi @troglobit,

I've done some first steps to build my own running linux system based on finit and busybox / alpine. Looks really great and works fine so far. 
Great job with finit 👍 

Now I thinking about rootfs and found an example to mount "/" from `/etc/fstab`. Is it something like switch_root or just mount a partition to "/"?

My idea was to mount squashfs and a overlayfs to / or squashfs with some persistent additional /<directory> mounts.

Is there an example with fstab rootfs mount / switch_root or squashfs / overlayfs?

Many thanks!

**Comment by troglobit** [comment]
Thanks, that's very kind of you :)

About 10-15 years ago there was a pivot root implementation in Finit. Since then nobody's use that.  What works, and is in use today, is a basic `mount -a` which uses `/etc/fstab` by default.  We recently added an option to use a different fstab from the finit command line (kernel command line with the `--` separator from the boot loader).

I have setups with Alpine, Debian working with Finit, there use the native ext2 partitions.  However, for embedded systems I employ Buildroot and there's an example of using a squashfs root in https://github.com/troglobit/myLinux -- specifically, see https://github.com/troglobit/myLinux/blob/main/board/common/rootfs/etc/fstab.ro which is used for `make run` when an amd64 build is started in Qemu.  Another (very similar) example is https://github.com/westermo/netbox/blob/master/board/common/rootfs/etc/fstab which is always used. Both use a BusyBox mount helper to actually mount the things, but I think you'll get the gist of it from the fstab file.

**Comment by aanderse** [comment]
sorry to revive an old discussion but i was hoping for some guidance on the topic

warning: i am no expert on building an os from ground up so pardon my ignorance where applicable 😅

how does this work for a standard debian install (for example) which has an initramfs and wants to run `finit`? i have seen where `systemd` is run in an initramfs and has explicit switch root functionally so it can transition from initramfs to a "real" mounted root... but i don't understand how i can do this with `finit` without resorting to the solutions you posted above (but those solutions don't seem appropriate for a "proper" desktop for example)

my thought is that switch root functionality would be immensely useful to start `finit` have it do some basic startup to allow mounting of the root filesystem (to an alternate location) and then switch root to it... all before reading `/etc/finit.conf` which of course resides on the mounted root

am i missing something obvious or is it simply that `finit` isn't commonly used in the scenario i am imagining? 

any and all guidance is much appreciated, thank you 🙇‍♂️

**Reply by troglobit** [reply]
Well, one of the reasons there's no pivot/switch root in Finit is that I've never managed to fully grasp how they work :smile: [looking at it now](https://www.marcusfolkesson.se/blog/changing-the-root-of-your-linux-filesystem/) it doesn't look so scary, although I don't know fully how to integrate that support in Finit atm. (yet I'm definitely open to discuss it)

Another reason is that most initramfs systems I've tested Finit on run another init (BusyBox usually) inside the initramfs and the use `switch_root ... ` to call finit.

**Reply by aanderse** [reply]
> yet I'm definitely open to discuss it

amazing!

> another init (BusyBox usually) 

i have seen an initramfs built with `systemd` - and once used [dinit](https://github.com/davmac314/dinit) to build one and i can't over emphasize how it nice it was

 `finit` offers so much functionality (in addition to conditions and ordering!) that it would make such a nice tool to build an initramfs with

---

i am also not sure how this works and unfortunately my work with `dinit` was really just a leaning experience because after `dinit` had finished preparing the system i ended the process and used switch root to hand the system over to a different init system 😅 

but maybe simply calling switch root from within `finit` at the right point is enough... 🤔

**Reply by wkz** [reply]
(I am also an initramfs-n00b :smile:)

My understanding is that in this use-case, `finit` in the `initramfs` would only:
1. Run through runlevel `S` as usual, and then
2. Call `switch_root` rather than transition into the default runlevel (e.g. 2 or 3)

Right? Because you can't leave any services running across the `switch_root` boundary anyway. So the reason for using `finit` would "only" be to sequence work in `S`. Or am I missing something?

**Reply by aanderse** [reply]
yes, something like that... i'm not exactly sure of the details yes

i quickly hacked up the initramfs i built with `dinit` to use `finit` instead - i was able to run the plugins (`bootmisc`, `modules-load`, etc...) and a few of my own `hook-scripts`, followed by mounting `fstab` with no problem

my next problem will be how to get a proper root filesystem mounted... but there are conceivably services i could want to run in initramfs (like ssh, maybe i have an encrypted root?) that would shutdown before calling switch root - maybe initramfs run as a custom runlevel? :thinking:

---

ps. @troglobit from my initial impressions this is _fanstatic_ software! the simplicity + power of `runlevel` + `conditions` is a dream :heart:

**Reply by aanderse** [reply]
i thought about this some more and i think one option that appeals to me is in effect a full instance of `finit` with separate configuration (`/etc/finit.conf`, etc...) in my initramfs to run to completion, not shut down the system, then switch root into a new `finit` instance which will have its own `/etc/finit.conf` on the root filesystem - does that make sense at all? 

i am curious to see what @troglobit thinks would fit best into the software

thanks for any thoughts shared

**Reply by troglobit** [reply]
@aanderse glad you like it, we like it too! :relaxed:

I still don't fully understand your use-case (ssh + encrypted root).  At first I thought maybe you meant running all run/task in runlevel S, and if `switch_root` is set in the config call that instead of switching to runlevel 2.   However, I talked with @wkz a bit this morning and he posited that you may want to start up a limited system with networking, including SSH login (with public key auth) which a remote agent could automate login, decrypt a partition, and then perhaps call `initctl switch root`?

**Reply by aanderse** [reply]
> one option that appeals to me

i will emphasize "one option" and "me", clarifying that i just wanted to add some ideas in an effort to brainstorm - regardless of what you end up thinking is a good idea i appreciate you simply listening and considering _my_ thoughts regarding _your_ project - i want to be respectful of your time and effort 🙇‍♂️ 

---

i watched a [really cool talk](https://m.youtube.com/watch?v=X-2zfHnHfU0) where this guy was doing some neat things in initramfs (which he calls "stage 1"), just like @wkz had considered

the speaker in that linked video gave an example of a motivating case: "using systemd stage 1 to unlock drives with a TPM and Tailscale", though the [demo code](https://github.com/ElvishJerricco/stage1-tpm-tailscale) is a bit convoluted to those unfamiliar with `nix` lang

---

how do you see the switch root working?

**Reply by troglobit** [reply]
> i want to be respectful of your time and effort 🙇‍♂️

Much appreciated :pray: 

> i watched a [really cool talk](https://m.youtube.com/watch?v=X-2zfHnHfU0) where this guy was doing some neat things in initramfs (which he calls "stage 1"), just like @wkz had considered the speaker in that linked video gave an example of a motivating case: "using systemd stage 1 to unlock drives with a TPM and Tailscale"

Very interesting, thanks for sharing that!

> how do you see the switch root working?

There are many ways to do this, from what I can see, systemd supports this over at least D-Bus and [systemctl switch-root](https://www.freedesktop.org/software/systemd/man/latest/systemctl.html#switch-root%20ROOT%20INIT) and the latter should be fairly straight forward to implement, we can also have a peek at [switch-root.c](https://github.com/systemd/systemd/blob/main/src/shared/switch-root.c)

Please keep in mind, we don't have the massive feature set that systemd does (in part intentional), so the impressive [systemd-ask-password](https://www.freedesktop.org/software/systemd/man/latest/systemd-ask-password.html) feature mentioned in the talk is not something we have.  I'd like to say *yet*, because if there's a good use-case for it I see no reason for *not* carrying our own implementations of the same ideas, and considering it could be integrated with ssh made it very interesting.

**Reply by aanderse** [reply]
sounds good! 

> Please keep in mind, we don't have the massive feature set that systemd does (in part intentional)

yes, i think you're doing a wonderful job at striking a balance between easy to use/understand + capable feature set so please keep it up! 

how would plugins and events work? would events like filesystem mounted fire twice and trigger plugins multiple times? that sounds like it could be useful - or perhaps separate events to hook into?

---

if you think this is fairly straightforward to implement and are willing to i am very excited to do thorough testing for you because i am very interested in getting this running on an old laptop i have

thanks

**Reply by troglobit** [reply]
> > Please keep in mind, we don't have the massive feature set that systemd does (in part intentional)
> yes, i think you're doing a wonderful job at striking a balance between easy to use/understand + capable feature set so please keep it up!

Thanks, it's a struggle ... :relaxed: 

> how would plugins and events work? would events like filesystem mounted fire twice and trigger plugins multiple times? that sounds like it could be useful - or perhaps separate events to hook into?

The way I see it is that you have one finit in your initramfs that runs until someone/somescript calls `initctl swith-root /path` (optionally with a second argument like the corresponding `systemctl` call) -- this would trigger the first finit instance to invoke the kernel [switch_root(8)](https://man7.org/linux/man-pages/man8/switch_root.8.html) syscall, which effectively replaces the first finit with the second (or another init process).  They will both be PID 1, weirdly enough.

> if you think this is fairly straightforward to implement and are willing to i am very excited to do thorough testing for you because i am very interested in getting this running on an old laptop i have

Finit is one of my favorite personal projects, that is also used in other projects I care about, but these days I have very little time in general to spend on implementing requests from others, for free.  This particular feature is interesting but unfortunately not something I need myself at the moment.

**Reply by aanderse** [reply]
ah, understood - at least this discussion serves as a record if anyone else had the same question

thanks for your time and explanation :bowing_man:

**Comment by troglobit** [comment]
This discussion has a related issue, #74

**Comment by aanderse** [comment, marked as answer]
@troglobit - i asked my friend [claude](https://claude.ai) to take a look at what it would take to add initramfs functionality to `finit` and after a bit of iterating we came up with [this](https://github.com/finit-project/finit/compare/master...aanderse:finit:switch_root)

i'm still in the _highly_ experimental stages of playing with this branch, but i wanted to reach out and see if this seems reasonable to you - after going over the code a bit and running some `qemu` tests i _think_ i'm on the right track

if you end up with some extra time and can take a quick look i would appreciate it, but no rush at all... holidays and all, enjoy your time off :smile:

**Reply by troglobit** [reply]
This looks more than reasonable!  Very cool indeed!

**Reply by troglobit** [reply]
Merged to main in https://github.com/finit-project/finit/commit/1e93fc167158eeb73108754678dc44ac8d171177
